src/util/rocksdb/properties.rs

```diff
@@ -466,6 +467,228 @@ impl DecodeProperties for UserCollectedProperties {
+    ) -> OffsetHandles {
+        let mut res = OffsetHandles::new(num_of_properties);
+        for (k, handle) in index_handles.0 {
+            let mut handles = vec![0; num_of_properties];
+            handles[idx] = handle.offset;
+            res.insert(k, handles);
+        }
+        res
+    }
+
+    fn insert(&mut self, key: Vec<u8>, offsets: Vec<u64>) {
+        self.offsets.insert(key, offsets);
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(1024);
+        for (k, offsets) in &self.offsets {
+            buf.encode_u64(k.len() as u64).unwrap();
+            buf.extend(k);
+            for offset in offsets {
+                buf.encode_u64(*offset).unwrap();
+            }
+        }
+        buf
+    }
+
+    fn decode(mut buf: &[u8], size: usize) -> Result<OffsetHandles> {
+        let mut res = OffsetHandles::new(size);
+        while !buf.is_empty() {
+            let klen = number::decode_u64(&mut buf)?;
+            let mut k = vec![0; klen as usize];
+            buf.read_exact(&mut k)?;
+            let mut offsets = Vec::with_capacity(size);
+            for _ in 0..size {
+                offsets.push(number::decode_u64(&mut buf)?);
+            }
+            res.offsets.insert(k, offsets);
+        }
+        Ok(res)
+    }
+
+    fn get_approximate_distance_in_range(&self, idx: usize, start: &[u8], end: &[u8]) -> u64 {
+        assert!(start <= end);
+        assert!(idx < self.num_of_properties);
+        if start == end {
+            return 0;
+        }
+
+        let range = self.offsets.range::<[u8], _>((Unbounded, Included(start)));
+        let start_offset = match range.last() {
+            Some((_, v)) => v[idx],
+            None => 0,
+        };
+        let mut range = self.offsets.range::<[u8], _>((Included(end), Unbounded));
+        let end_offset = match range.next() {
+            Some((_, v)) => v[idx],
+            None => self.offsets.iter().last().map_or(0, |(_, v)| v[idx]),
+        };
+        if end_offset < start_offset {
+            panic!(
+                "start {:?} end {:?} start_offset {} end_offset {}",
+                start, end, start_offset, end_offset
+            );
+        }
+        end_offset - start_offset
+    }
+
+    pub fn smallest_key(&self) -> Option<Vec<u8>> {
+        self.offsets.iter().next().map(|(key, _)| key.clone())
+    }
+
+    pub fn largest_key(&self) -> Option<Vec<u8>> {
+        self.offsets.iter().last().map(|(key, _)| key.clone())
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct RangeProperties {
+    pub offset_handles: OffsetHandles,
+}
+
+const RANGE_SIZE_OFFSET: usize = 0;
+const RANGE_KEYS_OFFSET: usize = 1;
+
+impl RangeProperties {
+    pub fn encode(&self) -> UserProperties {
+        let mut props = UserProperties::new();
+        props.encode(PROP_RANGE_INDEX, self.offset_handles.encode());
+        props
+    }
+
+    pub fn try_decode<T: DecodeProperties>(props: &T) -> Result<RangeProperties> {
```

Comment:
I think we can just call it "decode" because the caller doesn't need to know what it is trying.